File: Part1_강의자료/Ch05_정렬/문제별코드/1431_시리얼번호/1431.java
import java.util.Arrays;
import java.util.Scanner;

class Serial {
    public Serial(String serial) {
        this.serial = serial;
        for (int i = 0; i < serial.length(); i++)
            if ('0' <= serial.charAt(i) && serial.charAt(i) <= '9')
                sum += serial.charAt(i) - '0';
    }
    String serial;
    int sum;
}

class Main
{
    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        Serial[] serials = new Serial[N];
        for (int i = 0; i < N; i++)
            serials[i] = new Serial(sc.next());

        Arrays.sort(serials, (o1, o2) -> {
            if (o1.serial.length() != o2.serial.length())
                return o1.serial.length() - o2.serial.length();
            if (o1.sum != o2.sum)
                return o1.sum - o2.sum;
            return o1.serial.compareTo(o2.serial);
        });

        for (Serial s : serials)
            System.out.println(s.serial);
    }
}

File: Part1_강의자료/Ch05_정렬/문제별코드/1431_시리얼번호/1431.py
class Serial:
    def __init__(self, serial: str, sum: int):
        self.serial = serial
        self.sum = sum

N = int(input())
serials = []
for i in range(N):
    serial = input().strip()
    total = sum(int(c) for c in serial if c.isdigit())
    serials.append(Serial(serial, total))

serials.sort(key=lambda s: (len(s.serial), s.sum, s.serial))

for s in serials:
    print(s.serial)
